docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+that we mentioned in the previous section. The main differences are the following:
+
+1. Gradients from each worker are not sent to master.
+1. The average of gradients across all workers is calculated locally on each worker.
+1. The model is updated directly on each worker and each worker has the exact same copy of the model.
+
+Below is the pseudo-code for this process on each worker:
+
+```python
+communicator = AllreduceCommunicator()
+
+with tf.GradientTape() as tape:
+    outputs = model(features, training=True)
+    loss = loss_fn(outputs, labels)
+local_grads = tape.gradient(loss, get_trainable_items())
+res, averaged_grads = communicator.average_gradients(grads)
+if res == SUCCESS:
+    update_model(averaged_grads)
+else:
+    report_failure()
+    continue
+```
+
+#### Failure Handling during Training
+
+The above pseudo-code will be wrapped and executed for each batch of the dataset inside `process_minibatch_and_report()`
+in the code below. Each worker continues to perform tasks until there is no new batch available.
+
+```python
+while True:
+    dataset = self._task_data_service.get_dataset()
+    if not dataset:
+        break
+    dataset = dataset.batch(self._minibatch_size).prefetch(1)
+    for dataset_batch in dataset:
+        task = self._task_data_service.get_current_task()
+        process_minibatch_and_report(dataset_batch, task)
+```
+
+If any of the gradients averaging operation fails while the workers are still active and healthy, we simply report the
+failure and continue training on the next batch.
+
+If any of the workers fails, e.g. the pod is accidentally killed, the following steps will be performed:
+
+1. The task that the failed worker was handling will be added back to the task queue.
+1. Since the `AllreduceCommunicator` is aware of the failure. It will reconstruct the communicator and re-assign ranks among
+the existing active workers. The existing workers will then continue to run on the tasks at hand.
+1. The master pod will try to create a new worker pod after a specified restart delay period.
+1. Once the worker pod becomes active and `AllreduceCommunicator` is aware of it, we perform the following steps:
+    1. Lock all existing worker pods.
+    1. Select one of the worker pods to broadcast the current model parameters to all worker pods.
+    1. Release the lock that each worker pod holds.
+    1. All worker pods continue to process training tasks at hand and perform allreduce operations.
+
+Note that since the existing active workers have the exact same copy of the model after the previous allreduce operation completes,
+we can guarantee that the new worker will have the same copy of the model as the ones on other workers once the next
+allreduce operation completes.
+
+#### Training with Evaluation
+
+If the worker encounters any evaluation tasks in the above process, it will evaluate the model directly on the worker once the
```

Comment:
If we have N worker pods, M worker pods receive the evaluation task. The other N-M worker are training, will these worker pods wait in "communicator.average_gradients(grads)" until the M worker pods complete evaluation tasks?

Reply:
Yes since that's also what was originally designed in `model_evaluation.md`.

Reply:
@brightcoder01 FYI I added this to "Potential Future Optimizations" section.